protocol/linkpreview/linkpreview.go

```diff
@@ -241,13 +245,105 @@ func (u OpenGraphUnfurler) unfurl() (common.LinkPreview, error) {
 	return preview, nil
 }
 
+type ImageUnfurler struct {
+	url        *neturl.URL
+	logger     *zap.Logger
+	httpClient http.Client
+}
+
+func compressImage(imgBytes []byte) ([]byte, error) {
+	smallest := imgBytes
+
+	img, err := images.DecodeImageData(imgBytes, bytes.NewReader(imgBytes))
+	if err != nil {
+		return nil, err
+	}
+
+	compressed := bytes.NewBuffer([]byte{})
+	err = images.CompressToFileLimits(compressed, img, images.DefaultBounds)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(compressed.Bytes()) < len(smallest) {
+		smallest = compressed.Bytes()
+	}
+
+	if len(smallest) > maxImageSize {
+		return nil, errors.New("image too large")
+	}
+
+	return smallest, nil
+}
+
+func (u ImageUnfurler) unfurl() (common.LinkPreview, error) {
+	preview := newDefaultLinkPreview(u.url)
+	preview.Type = protobuf.UnfurledLink_IMAGE
+
+	headers := map[string]string{"user-agent": headerUserAgent}
+	imgBytes, err := fetchBody(u.logger, u.httpClient, u.url.String(), headers)
+	if err != nil {
+		return preview, err
+	}
+
+	if !isSupportedImage(imgBytes) {
+		return preview, fmt.Errorf("unsupported image type url='%s'", u.url.String())
+	}
+
+	compressedBytes, err := compressImage(imgBytes)
+	if err != nil {
+		return preview, fmt.Errorf("failed to compress image url='%s': %w", u.url.String(), err)
+	}
+
+	width, height, err := images.GetImageDimensions(compressedBytes)
+	if err != nil {
+		return preview, fmt.Errorf("could not get image dimensions url='%s': %w", u.url.String(), err)
+	}
+
+	dataURI, err := images.GetPayloadDataURI(compressedBytes)
+	if err != nil {
+		return preview, fmt.Errorf("could not build data URI url='%s': %w", u.url.String(), err)
+	}
+
+	preview.Thumbnail.Width = width
+	preview.Thumbnail.Height = height
+	preview.Thumbnail.DataURI = dataURI
+
+	return preview, nil
+}
+
 func normalizeHostname(hostname string) string {
 	hostname = strings.ToLower(hostname)
 	re := regexp.MustCompile(`^www\.(.*)$`)
 	return re.ReplaceAllString(hostname, "$1")
 }
 
+// isSupportedImageContentType detects whether a URL ends with one of the
+// supported image extensions. It provides a quick way to identify whether URLs
+// should be unfurled as images without needing to retrieve the full response
+// body first.
+func isSupportedImageURL(url *neturl.URL) bool {
+	re := regexp.MustCompile(`(?i)^.+(png|jpg|jpeg|webp)$`)
```

Comment:
Worth considering not using a regex here and compare char by char maybe? I run a few benchmarks and it would be around 10/14 times faster

```
func isSupportedImageURL(url *url.URL) bool {
	path := url.Path
	l := len(path)
	
	return (l > 5 && lower(path[l-1]) == 'g' && lower(path[l-2]) == 'n' && lower(path[l-3]) == 'p' && lower(path[l-4]) == '.') ||
		(l > 4 && lower(path[l-1]) == 'g' && lower(path[l-2]) == 'e' && lower(path[l-3]) == 'p' && lower(path[l-4]) == 'j' && lower(path[l-5]) == '.') ||
		(l > 5 && lower(path[l-1]) == 'g' && lower(path[l-2]) == 'e' && lower(path[l-3]) == 'p' && lower(path[l-4]) == 'j' && lower(path[l-5]) == '.') ||
		(l > 5 && lower(path[l-1]) == 'p' && lower(path[l-2]) == 'b' && lower(path[l-3]) == 'e' && lower(path[l-4]) == 'w' && lower(path[l-5]) == '.') ||
		(l > 5 && lower(path[l-1]) == 'p' && lower(path[l-2]) == 'b' && lower(path[l-3]) == 'e' && lower(path[l-4]) == 'w' && lower(path[l-5]) == '.') ||
		(l > 5 && lower(path[l-1]) == 'o' && lower(path[l-2]) == 'b' && lower(path[l-3]) == 'e' && lower(path[l-4]) == 'w' && lower(path[l-5]) == '.') ||
		(l > 5 && lower(path[l-1]) == 'a' && lower(path[l-2]) == 'b' && lower(path[l-3]) == 'e' && lower(path[l-4]) == 'w' && lower(path[l-5]) == '.') ||
		(l > 5 && lower(path[l-1]) == 'f' && lower(path[l-2]) == 'b' && lower(path[l-3]) == 'e' && lower(path[l-4]) == 'w' && lower(path[l-5]) == '.')
}

func lower(c byte) byte {
	if c >= 'A' && c <= 'Z' {
		return c + ('a' - 'A')
	}
	return c
}
``` 

(chatgpt generated)
a bit ugly, but could be worth it if used many times and if it's a vector of attack

Reply:
@cammellos, I think this level of performance is fine, because unfurling needs to happen every few hundred milliseconds at most (debounced), since the slowest path is IO bound, so the savings from this check should be negligible. Now that you mention about performance, I think I can optimize it enough by avoiding the recompilation of the regex all the time, so I can extract the compiled regex to a var.

About the security aspect, I understand historically regexps have been the source of some issues (not sure about Go's history), but if I can't use a regexp here, are you saying that in Go we shouldn't use regular expressions if the input string is out of our control?

Reply:
> About the security aspect, I understand historically regexps have been the source of some issues (not sure about Go's history), but if I can't use a regexp here, are you saying that in Go we shouldn't use regular expressions if the input string is out of our control?

In general, we try to avoid it, as performance can vary and it's possible at times to find a weak spot in a regex to dos the app. (also input in our case is when it comes from the network, not from the user using the app)
This is probably not the case, I think you'll be the best judge at that since you'll know the usage better, and seems like the regex is probably ok

Reply:
Thanks for the feedback, made me think for sure. For this particular case, the input string are URLs that already went through the URL parser, and, for the majority of time, have a reasonably small length. So with that in mind I'll keep the simpler version @cammellos, but I'll extract the regexp compilation to help a little bit.

Reply:
Done